### ISSUE TITLE
Thunderbolt controller firmware update - X1 Carbon gen6

### ISSUE THREAD

**Issue by LordChaos73** [fwupd/firmware-lenovo#264]
Hi,

I'm one of the many owners with this laptop and I've lost HDMI output. I'm not sure when it exactly happened, but I think it was when I updated the Intel ME firmware a month or three ago. Basically, nothing happens when an external monitor is connected.
I've read a lot about this particular issue on (Lenovo) forums and it seems the following firmware update can fix this issue:

https://pcsupport.lenovo.com/nl/en/products/LAPTOPS-AND-NETBOOKS/THINKPAD-X-SERIES-LAPTOPS/THINKPAD-X1-CARBON-6TH-GEN-TYPE-20KH-20KG/20KH/downloads/DS502615

The laptop has the latest BIOS update (1.57) and the latest Intel ME firmware installed as well.
The thing is, my laptop is running Arch Linux solely. Could this update be made available for Linux users (fwupdmgr) as well?

**Comment by hughsie** [comment]
@LordChaos73 what firmware version are  missing? Thanks.

**Comment by LordChaos73** [comment]
@hughsie My question is: is the thunderbolt controller firmware update (as linked above) available on the LVFS platform?

**Comment by mrhpearson** [comment]
Hmmm - looks like we last updated the TBT controller on the X1 Carbon 6 two years ago.. with version 43.
I'll check with the FW team when (hopefully not if) the next update is - it looks like they're now up to version 47.

All that being said - I don't see why a TBT controller update would have any impact on HDMI display - unless you're connected via a TBT dock maybe? Any reason you think this particular FW update will help? I'm happy to chase an update (it should be done) but I don't think it's going to fix the root cause of your issue.

I've created internal ticket  LO-1984 for tracking

**Comment by LordChaos73** [comment]
@mrhpearson please see the following posts:

https://forums.lenovo.com/t5/ThinkPad-X-Series-Laptops/X1-Carbon-6th-gen-HDMI-Thunderbolt-stopped-working/m-p/5120432

https://forums.lenovo.com/t5/thinkpad-x-series-laptops/x1-carbon-6th-gen-hdmi-not-working/td-p/4091511

https://forums.lenovo.com/t5/ThinkPad-X-Series-Laptops/X1-Carbon-6th-Gen-HDMI-Not-Working-and-Thunderbolt-Missing/m-p/4326149

**Comment by mrhpearson** [comment]
Got it - thanks.

**Comment by dal00** [comment]
I have the same issue with thunderbolt not working and hdmi also having issues. I'm even unable to see thunderbolt in lspci.
Windows can't see it in the device manager too and I'm unable to flash it from windows. 

@LordChaos73 Are you even able to see the thunderbolt controller in lspci?

**Comment by mrhpearson** [comment]
Do you know which version of FW you had on there previously? There was a critical update a while back (2019) that was published and I'm pretty sure we put those updates on LVFS.
https://support.lenovo.com/ca/en/solutions/ht508988-critical-intel-thunderbolt-software-and-firmware-updates-thinkpad

My understanding was that if the FW wasn't updated it was possible to brick your TBT controller.
If you already had that update then we can ignore that as a theory. If it's not working in Windows it is worth calling customer support for assistance - you may have a HW issue.

**Comment by LordChaos73** [comment]
I remember about the critical update back in 2019 and I've always had the latest possible firmware installed since. Since so many X1C6 owners seem to have this issue, I'm wondering if the patch was successful in the first place.
Anyway, since this was my production machine until very recently, I know everything was working fine until we had the Intel ME issues that failed to install. I think I lost HDMI output around that time. And if I remember correctly, there was also a new BIOS release at the time.
I'm also not seeing my Thunderbolt controller using lspci (I've attached the output of 'sudo lspci -v').
[lscpi.txt](https://github.com/fwupd/firmware-lenovo/files/9515388/lscpi.txt)

**Comment by dal00** [comment]
@mrhpearson I'm not sure as I'm unable to connect to the controller in any way that I know of.

But here is a dump of the fw from the W25Q80DVSNIG
[x1c6_tb3.zip](https://github.com/fwupd/firmware-lenovo/files/9517570/x1c6_tb3.zip)

**Comment by mrhpearson** [comment]
Ack - thanks for the details and they've been forwarded to the FW team.

**Comment by mrhpearson** [comment]
Hi - can you confirm the BIOS, EC and ME FW versions please (easiest is to grab them from the BIOS setup screen - F1 during early boot). The note from the FW team says:

_I checked history, there was an issue that system will lost TBT controller and fixed by one EC firmware on X1 Carbon G8 project, never hear this issue on X1 Carbon G6 project. So that is why I suggest to have a try with upgrade BIOS/EC to newest, feel sorry that cannot give specifics of which version of EC._

Not sure why they can't tell me which version of FW...(I'm pretty annoyed about it, but it gets worse)

It looks like they released a new BIOS/EC combo update on the Lenovo support site - **but the EC has not been released on LVFS** (the BIOS has). I'm really grumpy about that (the BIOS is updated)

Can you try the cab file that is on the Lenovo site (https://download.lenovo.com/pccbbs/mobiles/n2wul23w.zip). I did the update on mine ('fwupdmgr install N2WHT20W.cab' - I had to set 'OnlyTrusted=false' in /etc/fwupd/daemon.conf) and it looks good - but I also wasn't having TBT issues.

I don't know if the new EC will fix the issue - there is nothing in the internal release notes that references it - but it sounds like it should be tried. I'm pushing them to get the EC on LVFS ASAP.

Mark

**Comment by LordChaos73** [comment]
![x1g6_setup](https://user-images.githubusercontent.com/7850883/189748947-3d12cb9f-ca51-403b-ac27-734f1c6deb53.jpg)

I´m not sure the cab files you provided are the correct ones?

[levid@x1carbon6 n2wul23w]$ fwupdmgr install N2WHT20W.cab
Decompressing…           [ -                                     ]
No supported devices found

[levid@x1carbon6 n2wul23w]$ fwupdmgr install N2WET37W.cab
Decompressing…           [    \                                  ]
No supported devices found

**Comment by LordChaos73** [comment]
By the way, I downloaded the ISO of the latest BIOS and wanted to see if I could re-apply it. So I created a bootable USB stick with the latest BIOS, but I can't seem to re-apply it; the setup program states there is nothing to do. Is there any way to re-apply the BIOS update? I was hoping the ISO _is_ carrying the EC update as well.

**Comment by mrhpearson** [comment]
Sorry - those are X1Carbon 8 items...my bad.
I'll look for the X1 Carbon 6 ones.

**Comment by LordChaos73** [comment]
Hi @mrhpearson were you able to find them?

**Comment by mrhpearson** [comment]
I checked and the X1 Carbon 6 versions I can find are the same you already have - so I went back to the FW team and asked for any other suggestions. Not heard back yet.

**Comment by LordChaos73** [comment]
@mrhpearson what about the EC update that was never provided on lvfs?

**Comment by mrhpearson** [comment]
EC 1.22 is here (and under stable) https://fwupd.org/lvfs/firmware/5728

Checking the released notes on the support site it's the latest version:
   https://download.lenovo.com/pccbbs/mobiles/n23ul35w.txt:
   1.57      (N23UL35W)  1.57 (N23ET82W)      1.22 (N23HT37W)   01    2022/08/25

**Comment by LordChaos73** [comment]
Is there any hope on restoring Thunderbolt/HDMI functionality? A BIOS update maybe?

**Comment by LordChaos73** [comment]
I see there's a new version of the Intel ME available - 11.8.93.4323. It's not on LVFS yet?

**Comment by cdrn** [comment]
Howdy all -- having the exact same issue w/ the exact same bios. Any suggestions here?

**Comment by kmauleon** [comment]
> I see there's a new version of the Intel ME available - 11.8.93.4323. It's not on LVFS yet?

hi @LordChaos73 still under testing... hopefully will be available by end of this month... please open separate issue for proper tracking. thank you very much

**Comment by RAY5D** [comment]
@cdrn Same issue here. I'm trying the full hardware route by buying W25Q80DVSNIG chips to replace the one on my motherboard that has gone bad, and programming them through a CH341-based programmer. Before the fix, the laptop was not shutting down properly by Windows, could only pull 15W from the charger instead of 65W, and had no HDMI output. The binary file provided by @dal00 fixed the charging issue, but nothing else. I'm not sure if that's a compatibility issue between BIOS, EC, and TB versions or there are other problems. I have BIOS 1.57 and EC 1.22, which I think is the latest. I just need to find a working binary for the TB controller.

**Comment by tw0flower** [comment]
I also have this issue (HDMI and DP not working). When I first noticed it I updated all firwmares through fwupd so I have everything on the latest version now. I have no idea if the critical fix from 2019 was applied on this computer as it is second-hand so I'm not very hopeful in my case. Does anyone know if it was rolled out automatically through Windows Update?

I'm in the process of getting an EEPROM programmer myself, because the controller does not even appear in lspci output or in Windows Device Manager. 

Do you still have the original data that was on your chip before you flashed it by any chance @RAY5D ? I'd be interested in comparing it to what I have. It's interesting that after changing the chip for a new one you still seem to face some issues.

**Comment by RAY5D** [comment]
@tw0flower Sorry, I don't have any data backed up. When I used the programmer on the old chip, it read all 0xff, which I think is the "default" value for this type of storage. It will still read 0xff even after writing whatever data, so the chip is most likely dead as dust since the programmer could still read its model number, indicating a correct connection. I'm pretty sure the only thing I need now is the binary file, but that is, unexpectedly, hard to find. Can you post your data here so I can try? Thanks!

**Comment by tw0flower** [comment]
@RAY5D Ah, I misunderstood your comment. I thought you had already replaced the chip, and that some other component was writing 0xFF data all over the new chip after you flashed it.

I will post my data, but I don't think flashing it will be useful as my computer also has the HDMI/DP problem. Have you checked these binary files on the LVFS website? ﻿https://fwupd.org/lvfs/devices/com.lenovo.ThinkPadN23TF.firmware

During my desperate search for answers on this issue, I also found [this blog post](https://nadim.computer/posts/2018-10-26-thinkpad.html) from a user that had a somewhat similar problem years ago on a Thinkpad P1. He apparently had to flash zeroes over the chip before being able to flash the actual firmware.

It goes without saying, I don't know much about hardware so anyone reading this please take it with a grain of salt. At this point I'm desperate and accepting the risk of bricking things.

**Comment by RAY5D** [comment]
@tw0flower I did replace the chip and flash the firmware. I also saw that post earlier, but apparently I didn't notice the part about writing 0's first and then flashing the actual firmware (oops). I'll do more testing with the binary files I found and post updates here. However one thing that confused me was that the official binary files from both Lenovo and https://fwupd.org/lvfs/devices/com.lenovo.ThinkPadN23TF.firmware are exactly 260KB while the chip is 1MB, and the dumped binary in this issue has some content after that region, so I'm not sure about what should I write after the 260KB.

**Comment by tw0flower** [comment]
@RAY5D That's what I hope I will find out if I can still extract some data from my chip. There might be some padding involved. I wonder if this information might not be available somewhere in the LVFS files? Because the flashing software must be aware of it right?

But in theory if you zero-out the chip first, then flash the firmware, this would ensure that only the firmware on the chip I think. If there is some padding involved you could also try writing it at the end of the memory by concatenating some zeroes followed by the 260kb file so that the total is 1MB. If the padding is some arbitrary value however we have to find another way.

**Comment by RAY5D** [comment]
@tw0flower OK I did some more testing. I had 2 binary files, the first one was the 260KB official release, with the extra space filled with 0's, the second one is the dumped data from this issue posted by [dal00](https://github.com/dal00). For each test case, I filled the chip with a certain value (0x00 or 0xff), started the PC, shutdown, flashed the binary, and start again to see if it works. Here are the results: 

1. 0x00 fill, 260KB+0's: No HDMI, charging works, Windows couldn't shutdown properly. 
2. 0xff fill, 260KB+0's: No HDMI, charging works, Windows couldn't shutdown properly. 
3. 0x00 fill, dumped: No HDMI, charging works, Windows couldn't shutdown, sometimes freezes and goes into BSOD. 
4. 0xff fill, dumped: Windows did not complete boot, black screen at login, tried Manjaro, no HDMI. 

The results looked overall similar to my previous attempts. Only the charging speed was fixed. However, before the tests today, I noticed that the replacement chip was dead again (wtf!?!?). I think I tried enabling assist mode in BIOS earlier, maybe that caused the early death? Anyways, I just replaced the chip again before the tests since I bought 10, and was careful to only enable assist mode while using Linux, and the chip still works for now. I don't think there is a lot left to do before we can find more binaries to compare and test with. Should I try the official binary but extended with 0xff next time?

**Comment by tw0flower** [comment]
I got my CH341 SPI today and I found a few things. 

First of all @RAY5D if you only see 0xFF bytes on your chip it might be that your clamp is not properly attached. I had this issue at first, I solved it by re-attaching it until I read some data and by using flashrom instead of ch341eeprom. If the RUN led on your SPI board is not on when reading/writing it means something is wrong in your setup.

The first thing I did was to extract the firmware I had installed on my machine. Compared to the firmwares I found online, it had an offset: the first 4096 bytes were all 0xFF. What remained was similar to the firmares I found online.

My initial situation is that everything worked except plugging a second display (HDMI or via Thunderbolt).
I then tried the following:
1.  I flashed the newest 47 firmware from the LVFS website. No changes.
2. I flashed the 47 firmware with a 4096 bytes offset. Charging was not using the full power from the adapter. 
3. I flashed the 43 firmware from the LVFS website. Charging was working normally again but no luck with the screen.
4. I went in the BIOS and reset the options to default with Windows 10 optimised settings. But unfortunately that didn't help.

So in my case:
- My problem is not caused by a hardware defect from the chip that stores the Thunderbolt firmware. I can read and write on it perfectly. I even made sure that reading after writing a file produced the exact same result.
- Changing the firmware does have an effect. I lost charging. But it had no effect on the external display.
- There was an offset on my firmware. I wasn't compatible with the 47 version. I have not tried it on the 43 version, but it made charging work again without it.

The files that I flashed on the chip are in the attached zip file.
[firmwares_x1cgen6.zip](https://github.com/fwupd/firmware-lenovo/files/10074138/firmwares_x1cgen6.zip)

**Comment by RAY5D** [comment]
@tw0flower Pretty sure I'm indeed getting dead chips since I always check if the programmer can read the chip's model number as an indication of good connection. However, when I tried your binary, the result is basically the same as my previous attempts. Only charging works, still no HDMI, and Windows still couldn't shutdown the laptop properly. Also, the binary file from [dal00](https://github.com/dal00) has something in the first 4096B. I'm not sure why yours are just 0xff in that region, it looks like those binaries are more complex than the flash-and-done type of job I thought it would be... The only thing I can think of right now is to get a TB dock or something, so I can "activate" the TB controller by plugging it in. That way I can try to update the firmware through the OS. However, this will probably not work given that the controller is barely working. Did you try the binary in this reply ↓↓↓?

> @mrhpearson I'm not sure as I'm unable to connect to the controller in any way that I know of.
> 
> But here is a dump of the fw from the W25Q80DVSNIG [x1c6_tb3.zip](https://github.com/fwupd/firmware-lenovo/files/9517570/x1c6_tb3.zip)

**Comment by jnada2021** [comment]
Hello, I am the person who opened the case with Lenovo forums (the one in the first message on this thread) and I also have a post in reddit about this issue dated last year. 

I have the latest BIOS 1.58 firmware which I flashed today. Latest Intel ME Firmware and software installed.
I flashed all the new TB firmwares posted in this forum using my  CH341 SPI with AS programmer. 
As with everyone, only the charging works and also with a USBC hub or USBC to USB A adapter I can connect basic USB devices (such as mouse etc.) but no Display from TB/USBC ports nor HDMI.
If it's any useful, here are the collection of all of the original Lenovo TBT firmwares from 12 to 24

[tbt12_23.zip](https://github.com/fwupd/firmware-lenovo/files/10096074/tbt12_23.zip)

I also have historical archive of X1 6C bioses starting with 19 to 31 which I had downloaded from Lenovo website. I can't upload to this post due to 161MB filesize limit. 
If it is useful for someone (to roll back etc.) I can also upload to  google drive.

**Comment by jnada2021** [comment]
And here is the latest N23TF24W firmware bin dated 25 May 2022 which I flashed as 260kb (only charging and basic USB work-No DP, No HDMI still) and read the chip again and saved 1MB this time.
[20221126_N23TF24W.zip](https://github.com/fwupd/firmware-lenovo/files/10096105/20221126_N23TF24W.zip)
https://pcsupport.lenovo.com/us/en/products/laptops-and-netbooks/thinkpad-x-series-laptops/thinkpad-x1-carbon-6th-gen-type-20kh-20kg/downloads/driver-list/component?name=USB%20Device,%20FireWire,%20IEEE%201394,%20Thunderbolt

IIRC the original TB firmware had the same info written twice one in the first half and the second instance is the latter part. I am not sure if my bin was corrupted and had repeated twice or it is because we have 2 TB ports. But someting to think about (may be)

**Comment by dal00** [comment]
There used to be separate .bins for assisted mode and native mode. You can find them in the extracted firmwares this guy collected if I remember correctly. https://www.badcaps.net/forum/printthread.php?t=82556

On November 26, 2022 12:30:21 PM GMT+01:00, jnada2021 ***@***.***> wrote:
>And here is the latest N23TF24W firmware bin dated 25 May 2022 which I flashed as 260kb (only charging and basic USB work-No DP, No HDMI still) and read the chip again and saved 1MB this time.
>[20221126_N23TF24W.zip](https://github.com/fwupd/firmware-lenovo/files/10096105/20221126_N23TF24W.zip)
>https://pcsupport.lenovo.com/us/en/products/laptops-and-netbooks/thinkpad-x-series-laptops/thinkpad-x1-carbon-6th-gen-type-20kh-20kg/downloads/driver-list/component?name=USB%20Device,%20FireWire,%20IEEE%201394,%20Thunderbolt
>
>IIRC the original TB firmware had the same info written twice one in the first half and the second instance is the latter part. I am not sure if my bin was corrupted and had repeated twice or it is because we have 2 TB ports. But someting to think about (may be)
>
>
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/fwupd/firmware-lenovo/issues/264#issuecomment-1328029276
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by jnada2021** [comment]
>     * My problem is not caused by a hardware defect from the chip that stores the Thunderbolt firmware. I can read and write on it perfectly. I even made sure that reading after writing a file produced the exact same result.
> 
>     * Changing the firmware does have an effect. I lost charging. But it had no effect on the external display.
 

Same as you. I can read and write to the W25Q80DVSNIG chip with the CH341-based programmer as you mentioned. Erasing the chip (with AsProgrammer for example) or writing the "wrong" *.bin file disables the charging. But if I flash a TBT.bin from Lenovo etc. charging and basic USB functions are back (USB 2 mouse, USB harddisk, connect to Sony camera etc.) but no TB, no DP and most importantly and quite weirdly no HDMI at all. I also have 2 good quality USBC to HDMI hub/adapters both work with Surface Go etc. so they are not crap but when I connect them to TB port zero display. Yes the pass through charging and basic USB functions also work with these hubs. 

Nowadays as a last resort I am using a dinky USBA HDMI adapter which only has 1080p max output but this solution is also not ideal for me since at the slightest movement of the cable or laptop, HDMI disconnects and I have to reboot each time (due to the bad design-fault of the HDMI adapter). Also I get weird looks from collegues as well since instead of using built in HDMI, I am now using a dinky USBA HDMI thingy.

I think the intel ME firmware (biggest spyware-CIA backdoor of the history) messed up something in some other area and writing efforts to W25Q80DVSNIG chip mean nothing. I wish we all could "factory reset" the X1 C6 to go back to its 2018 out of factory state (with OS, BIOS etc.) so that we could start all over again.

Another area to look at :
Does anyone know in which chip Intel ME firmware is written at and we can also attack to that chip to erase/reset and try?

Lenovo has been very mum on this issue and I think they are deliberately downplaying it since perhaps only some limited motherboard-chip combinations are affected or it might be a much bigger issue that they play dumb and hope that it will go away.

**Comment by jnada2021** [comment]
Some updates.

1.
I think Lenovo's made some changes in their latest 1.58 BIOS firmware so that even if the TB chip is corrupted or "empty" charging function works.
I have just loaded a zero.bin file to erase the TB chip and rebooted to test, and charging still works. In the past it would not and would at least give the low power warning before and/or don't charge.
Here is the zero.bin file if you want to test:

[zero_erase.zip](https://github.com/fwupd/firmware-lenovo/files/10098020/zero_erase.zip)

2.
I opened my very first original .bin file dated 30/12/2021 with HxD hex editor and the 260kb info has been written on it twice. 
And the very first one starts at line 00004096:
![padded](https://user-images.githubusercontent.com/79317095/204130314-434f3f1e-bf0d-4688-b346-4a5f66cceb81.jpg)

And the second at line: 00532480:
![2nd_repeat](https://user-images.githubusercontent.com/79317095/204130355-fa5002bc-dbc2-4e1e-9c2d-0cd5dd7df80e.jpg)


The more I look at this the more I think that the "corruption/problem" etc. is inside the intel ME chip/firmware whose contents probably are classified as "state secret" :) LOL

**Comment by RAY5D** [comment]
![20221110_154746](https://user-images.githubusercontent.com/12710440/204161067-bf94d05d-f873-4fe6-afa4-f9a1461fe60b.jpg)
@jnada2021 The BIOS and ME firmware is stored in another flash chip right above the CPU, near the fan connector. I think it's also readable with the same programmers for the TB chip. But I'm not sure how that has an affect on our issues...

**Comment by jnada2021** [comment]
IIRC the Lenovo "2019 fix" was to disable Intel ME writing something somewhere but even that was also inevitable. Remember  them saying "after applying the fix" don't ever update intel ME ever again, which makes me believe above conclusion. Unbenownst to many Intel ME is a "master OS" it runs concurrently with any computer regardless of it being Linux, Windows etc.

Were we able to extract any known good copy of the TBT bin file from someone who doesn't have TB, DP, HDMI problems yet? What dall00 shared in Sept 7th is also a "bad bin" since he extracted them after having the issue.

**Comment by RAY5D** [comment]
@jnada2021 Huh, that's interesting. Can you still find the article that mentioned it? Also, even if we manage to find someone with a fully working computer and extract both the ME and TB firmware, is it as easy as just flashing it into our chips? I would assume there are some bullshit encryption going on that binds the binary with a single motherboard or something.

**Comment by jnada2021** [comment]
1. These are the articles:
https://www.notebookcheck.net/ThinkPad-Thunderbolt-3-failure-What-s-happening-why-it-s-happening-and-how-to-fix-it.451207.0.html
https://www.reddit.com/r/thinkpad/comments/eszg2o/thinkpad_tb3_failure_what_we_know_repair/

Search: DO NOT FORCE UPDATE Thunderbolt Controller. This may damage the firmware.

2. By finding a "working bin" we at least eliminate one (important) possibility I am hoping. I am not sure of any encyription etc. since this guy was able to flash it with the programmer as we are trying but his machine is not a X1.
https://nadim.computer/posts/2018-10-26-thinkpad.html


According to below guide ME is also flashable &  editable. I may try this next weekend but I need to do more research on it since I don't know which chip is it on the MB and consequences of playing with that .... 

I guess it's better if we have a working copy of TBT firmware first and then (if that doesn't work) try this below:
https://github.com/corna/me_cleaner/wiki/External-flashing

**Comment by jnada2021** [comment]
Oh BTW if any of you are in US:
https://chimicles.com/lenovo-thinkpad-usb-c-charging-defect-class-action-investigation/

**Comment by LordChaos73** [comment]
I've always had my laptop on the latest (lvfs) updates; I haven't ran Windows in years. Still, I've lost HDMI/TB3 output a few months ago.
I'd like to thank @jnada2021 and @RAY5D for all their hard work regarding this issue.
Where is Lenovo in all of this? I have a (out of warranty) laptop I no longer can use at work since I no longer can connect it to an external display. I'd like to know if this issue is fixable through some firmware updates. Or should I buy a new machine?

**Comment by tw0flower** [comment]
Thank you everyone for looking in this issue. It seems a lot of people are affected by this. What bugs me is that I am pretty sure that I did not update anything before the Thunderbolt connector stopped working. I then did some updates to try and fix it, but originally I did not do anything at all. This is a very wild guess, but could it be caused by a certificate expiring somewhere?

I need to do more research on Thunderbolt but from what I could gather from [this article](https://www.dell.com/community/Latitude/Demystifying-Thunderbolt-3-Security-Levels/td-p/7529712), there are 4 modes to Thunderbolt connectors:
- USB 2.0
- USB 3.x
- DisplayPort
- PCIe

It seems to me that we are locked in USB-only mode whereas we should be working in PCIe mode with our USB/display adapters (which are really miniaturized docks).

I have also confirmed that switching modes and enabling/disabling BIOS assist mode does not change anything to the content of the firmware chip.

I'd like to look into more details in how the firmwares work and how they are supposed to be verified by the BIOS or Intel ME. I will keep you updated if I make any progress.

**Comment by tw0flower** [comment]
I did not have time to work directly on the firmwares that we have, but I searched Google again for information about Thunderbolt. I think these slides from a 2020 Black Hat conference could be useful: https://i.blackhat.com/USA-20/Thursday/us-20-Ruytenberg-When-Lightning-Strikes-Thrice-Breaking-Thunderbolt-3-Security.pdf

They describe the Thunderbolt architecture in detail with some hardware examples that are similar to the X1 Carbon that we have. And the different sections of the firmwares.

I'm attaching the PDF to this comment for safekeeping.
[us-20-Ruytenberg-When-Lightning-Strikes-Thrice-Breaking-Thunderbolt-3-Security.pdf](https://github.com/fwupd/firmware-lenovo/files/10128282/us-20-Ruytenberg-When-Lightning-Strikes-Thrice-Breaking-Thunderbolt-3-Security.pdf)

**Comment by dal00** [comment]
I found a tool referenced in that pdf https://github.com/BjornRuytenberg/tcfp which looks like can parse the firmware.

When running it on TBT.bin extracted from n23th08w I get the following output:
I wonder if we're missing something important since it says the image is incremental.

```
$ ./tcfp.py  parse TBT.bin 
2022-12-01 11:25:25,998 - WARNING - File size in between 229376 and 1048576 bytes. Possible causes: 
 - Image may be an incremental firmware update. While tcfp may be able to parse the SL state, please note patching requires a full dump. 
 - Image dump may be incomplete, i.e. not include 'scratch pad' section. However, this should typically not cause any issues.
Vendor ID : 0x109
PCI ID : 0x15d3
PCI Device Name : JHL6540 Thunderbolt 3 Bridge (C step) [Alpine Ridge 4C 2016]
Model ID : 0x1706
NVM version : 47 (0x2f)
Vendor : Lenovo
Device : X1 Carbon
Image type : Incremental
Security Level : SL1
```

**Comment by kubabuda** [comment]
@dal00 

> TBT.bin extracted from n23th08w 

How to do it? When I just unzip TB firmware, there are multiple .bss, .data, .edata etc sections, no memory map
Is there tool or commandline spell (linker?) to just get my .bin from that?
This is probably trivial, but I could not google anything useful

**Comment by jnada2021** [comment]
> @dal00
> 
> > TBT.bin extracted from n23th08w
> 
> How to do it? When I just unzip TB firmware, there are multiple .bss, .data, .edata etc sections, no memory map Is there tool or commandline spell (linker?) to just get my .bin from that? This is probably trivial, but I could not google anything useful


Are you using Linux or Windows 10/11. If you are using Windows it creates a folder under C drive:
C:\DRIVERS\WIN\ThunderboltFW
BTW here is the latest FW TBT file:
[TBT.zip](https://github.com/fwupd/firmware-lenovo/files/10164767/TBT.zip)

**Comment by kubabuda** [comment]
Thanks

**Comment by tw0flower** [comment]
I tried downgrading the BIOS of my computer yesterday. I really don't get why our Thunderbolt are working in USB mode but completely transparent at the same time (no listing in Windows Device Manager and lspci). Could it be a killswitch that was introduced in some an update?

The paper published on the Thunderspy vulnerability states that there is no fix for computers released before 2019. The X1 Carbon Gen 6 was released in 2018. So this vulnerability simply cannot be fixed when it comes to memory protection.
Therefore it's also unlikely that the BIOS or Intel ME could detect our firmwares as being "untrusted" at least that's what the author of the paper seems to suggest in the Q&A:

> [Why has Intel not fixed Thunderspy?](https://thunderspy.io/#intel-not-fixed-thunderspy)
> In our vulnerability disclosure procedure, Intel has confirmed the following:
> 
>     All three versions of Thunderbolt are affected by the Thunderspy vulnerabilities.
>     Only systems shipping Kernel DMA Protection mitigate some, not all, of the Thunderspy vulnerabilities.
>     Only some systems that began shipping since 2019 come with Kernel DMA Protection.
>     Beyond [Kernel DMA Protection](https://thunderspy.io/#kernel-dma-protection), Intel will not provide any mitigations to address the Thunderspy vulnerabilities. Hence, Intel will not assign any CVEs to the Thunderspy vulnerabilities, or release any public security advisories to inform the general public.
> 
> Despite our repeated efforts, the rationale to Intel's decision not to mitigate the Thunderspy vulnerabilities on in-market systems remains unknown. Given the nature of Thunderspy, however, we believe it would be reasonable to assume these cannot be fixed and require a silicon redesign. Indeed, for future systems implementing Thunderbolt technology, Intel has stated they will incorporate additional hardware protections.
> 
> Kernel DMA Protection is only available on a limited number of recent systems. Hence, all systems released before 2019, and more recent systems that do not ship Kernel DMA Protection, will remain fully vulnerable to Thunderspy forever. Please refer to Kernel DMA Protection for more information.

Thunderspy involves several problems. One of which is that the signature of the firmware is not checked during boot. I believe this is what the author is referring to when he says that Intel has not done anything beyond Kernel DMA Protection.

But what if they found a way to change this recently and introduced a bug? And the chip is wrongly and silently being deactivated at some stage?

Downgrading the BIOS did nothing, and we cannot downgrade Intel ME through software solutions, I've also tried. Might be doable through a SPI programmer.

I've begun analyzing the files I have with tcfp yesterday, I'll write my results here tonight.

**Comment by jnada2021** [comment]
> I tried downgrading the BIOS of my computer yesterday. I really don't get why our Thunderbolt are working in USB mode but completely transparent at the same time (no listing in Windows Device Manager and lspci). Could it be a killswitch that was introduced in some an update?
> 
> The paper published on the Thunderspy vulnerability states that there is no fix for computers released before 2019. The X1 Carbon Gen 6 was released in 2018. So this vulnerability simply cannot be fixed when it comes to memory protection. Therefore it's also unlikely that the BIOS or Intel ME could detect our firmwares as being "untrusted" at least that's what the author of the paper seems to suggest in the Q&A:
> 
> > [Why has Intel not fixed Thunderspy?](https://thunderspy.io/#intel-not-fixed-thunderspy)
> > In our vulnerability disclosure procedure, Intel has confirmed the following:
> > ```
> > All three versions of Thunderbolt are affected by the Thunderspy vulnerabilities.
> > Only systems shipping Kernel DMA Protection mitigate some, not all, of the Thunderspy vulnerabilities.
> > Only some systems that began shipping since 2019 come with Kernel DMA Protection.
> > Beyond [Kernel DMA Protection](https://thunderspy.io/#kernel-dma-protection), Intel will not provide any mitigations to address the Thunderspy vulnerabilities. Hence, Intel will not assign any CVEs to the Thunderspy vulnerabilities, or release any public security advisories to inform the general public.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Despite our repeated efforts, the rationale to Intel's decision not to mitigate the Thunderspy vulnerabilities on in-market systems remains unknown. Given the nature of Thunderspy, however, we believe it would be reasonable to assume these cannot be fixed and require a silicon redesign. Indeed, for future systems implementing Thunderbolt technology, Intel has stated they will incorporate additional hardware protections.
> > Kernel DMA Protection is only available on a limited number of recent systems. Hence, all systems released before 2019, and more recent systems that do not ship Kernel DMA Protection, will remain fully vulnerable to Thunderspy forever. Please refer to Kernel DMA Protection for more information.
> 
> Thunderspy involves several problems. One of which is that the signature of the firmware is not checked during boot. I believe this is what the author is referring to when he says that Intel has not done anything beyond Kernel DMA Protection.
> 
> But what if they found a way to change this recently and introduced a bug? And the chip is wrongly and silently being deactivated at some stage?
> 
> Downgrading the BIOS did nothing, and we cannot downgrade Intel ME through software solutions, I've also tried. Might be doable through a SPI programmer.
> 
> I've begun analyzing the files I have with tcfp yesterday, I'll write my results here tonight.

I hope your research bares fruit since so far you and dal000 have been the most resourceful on this issue.

I keep asking myself as to how to remove/roll back ME since the chip we have been writing the tbt.BIN to is not really corrupted, I think. I write zero.bin file and load any version of TBT.bin and boot and reconnect and read with the SPI clip again. Do a Beyond compare etc. and all versions of the bin files seem to be the same.

As of this week I have a brand new problem (after the BIOS update to 58 and consequently updating ME firmware to the latest version), no matter which TBT.bin I load (including very old ones that solved the low wattage charging problem), during boot and within the OS I have low wattage charging issue. It is neither the USBC cables, nor the port nor the power adapters since I have 2 original 65W usbc and also some other 3rd party (high quality) adapters.

To check if it were the USBC port or the cable problem, I disconnected the built in battery while in the OS and plugged AC in and it still kept working even reboot works but when I remove the battery and try to run on AC only from a full shutdown no dice. KB lits up, CPU fan spins for some seconds and then silence and no boot. Until of course I reconnect the battery and AC back things start working again.

I am very tempted to do remove as much ME and load the stripped BIOS back with SOIC clip but I fear I will brick the machine and render it completely useless. I have already extracted&downloaded the contents of the BIOS dump 4 times and all are the same according to beyond compare.
https://github.com/corna/me_cleaner/wiki/External-flashing

I loved thinkpads for over 20+ years but this issue I think is wearing me down and I will throw the towel and never touch a thinkpad ever again.

**Comment by tw0flower** [comment]
Everyone's research on this has been very useful IMO. We have no idea what's causing the issue and it takes a lot of time to dig information. And we're all doing that on our free time.

I'm personally not convinced that the problem lies with the Thunderbolt chip itself. But this might be wrong. One interesting thing to try would be to patch one of the bin files with the Thunderspy vulnerability to force it to the SL0 level.
Unfortunately, and this is what my attempts from yesterday showed, none of the firmwares I've tried has a signature compatible with the signatures provided by the tcfp tool. So we'd have to find how to patch it ourselves.

All images provided by Lenovo are using the SL1 level (user authorization). One interesting thing to note is that the version 47 advertises a fix for a vulnerability without mentioning which one. I don't know any other Thunderbolt-specific vulnerability outside of Thunderspy.

As opposed to the newer firmwares provided by Lenovo, my original firmware's SL level is not even detected by the tool. When I looked manually it seemed close to one of the signatures listed in the tool's code (see [here](https://github.com/BjornRuytenberg/tcfp/blob/master/tcfp.py#L49)) in terms of offset but the values are different. I have not tried any of the other firmwares that people uploaded here but it might be worth a try.

I'd like to contact the researcher that worked on the Thunderspy paper to ask him if he heard of any mitigation that might have been developed recently. I'll write him a message tomorrow or this weekend.

I've only ever owned Thinkpads as well, my first one was a T42 that I might still have somewhere haha! I remember a common fix for the ATI GPU overheating was to put the motherboard in the oven. Maybe we should try that ;-)

**Comment by jnada2021** [comment]
It being a holiday with heavy rains, I caved in and attempted to "disable" Intel ME b using below tutorial. As per the screen capture attached in BIOS ME doesnt report the version.
https://x86sec.com/posts/2022/09/08/removing-intel-ME-on-x1-carbon/

No change in the status of DP, HDMI or TB. Back to the drawing board.
![20221208_153002](https://user-images.githubusercontent.com/79317095/206475926-fc44ffd5-ee0b-447e-94a3-baf87ade0a14.jpg)

**Comment by jnada2021** [comment]
Is there a known way to get the whole dump of TBT.bin file 1MB inside the chip we have been flashing without the external clip? The fellow I used whose tutorial about disabling intel ME owns a 6th gen Carbon with i7 and is willing to share his file since his HDMI, DP and thunderbolt is working. Unfortunately he doesn't have access to the SOIC clip at the moment hence the question.
Cheers!

**Comment by jnada2021** [comment]
Hey. I've managed to get a working bin from a generous donor.
Give it a try and let me know if it solves your issues. Happy holidays!
[x1c6-win-tb.zip](https://github.com/fwupd/firmware-lenovo/files/10304602/x1c6-win-tb.zip)

**Comment by RAY5D** [comment]
> Hey. I've managed to get a working bin from a generous donor. Give it a try and let me know if it solves your issues. Happy holidays! [x1c6-win-tb.zip](https://github.com/fwupd/firmware-lenovo/files/10304602/x1c6-win-tb.zip)

Surprise Christmas present O.o?! Did you get it through a programmer or some sort of software means? My friend has an Thinkpad tablet on the problematic list and wants to back up the firmware before anything bad happens. However, the flash chip in that model is not readable without desoldering, and is a pretty annoying DFN package, so it would be very helpful if we can get the data without messing with the motherboard.

**Comment by jnada2021** [comment]
Some good samaritan responded to my post from December last year. He extracted it with a SOIC clip.

**Comment by RAY5D** [comment]
Still not working for me. If the bin works for other machines, the problem might be that it contains a public key as mentioned in the Black Hat Thunderbolt 3 Security slide page 29. I searched in the file, and the string "RSA" appeared at 0x38E00 and 0xB8000. The content seems to be the same at both locations, maybe the 2 copies are for assist mode and no assist mode? Anyways, page 31 says all TB3 controllers have unique IDs fused in silicon, but it also mentioned that the statement is inaccurate, not sure what that means. However if we need unique keys just to use the binary on our TB3 chips, we'll have to find a way to read the ID. Any ideas?
![image](https://user-images.githubusercontent.com/12710440/209585081-1beec886-902a-4daa-a06d-9434cbccb570.png)

**Comment by jnada2021** [comment]
Well, since many people in this post were using nothing but linux and never updated their Intel ME, I have a sneaky suspicion that the problem we have been experiencing may probably be a mix of both hardware and software fukups combined. Those 2 chips near the ROM chip that we have been flashing may be the culprit or some other BS issue that is injected via BIOS, Intel ME which burns/damages something somewhere.

My first USBC-Thunderbolt port was shot as of this summer (it was always finicky to begin with) and since a month or so my remaining USBC-TB port (the combination one) only charges at 15W no matter what I do (in terms of flashing the ROM chip). And I always used original Lenovo/Thinkpad USBC adapters and ports were never dinged etc, so go figure!

Lenovo folks finally made impossible possible and cracked the code of planned obsolescence. I feel bad for people who paid more than a grand for an X1 Carbon since let's be real, other than the keyboard and soft black coating the difference between mid/high range Lenovo consumer laptops and Carbon line is very negligable since 2017.

I think I will try to sell mine for a couple hundred euros i7-16GB, 1080p IPS and 6hr on battery and look for something else.

**Comment by andreidragu** [comment]
> Hey. I've managed to get a working bin from a generous donor. Give it a try and let me know if it solves your issues. Happy holidays! [x1c6-win-tb.zip](https://github.com/fwupd/firmware-lenovo/files/10304602/x1c6-win-tb.zip)

can you help me get one for T580?

**Comment by jclaveau** [comment]
@LordChaos73 
> I've always had my laptop on the latest (lvfs) updates; I haven't ran Windows in years. Still, I've lost HDMI/TB3 output a few months ago. I'd like to thank @jnada2021 and @RAY5D for all their hard work regarding this issue. Where is Lenovo in all of this? I have a (out of warranty) laptop I no longer can use at work since I no longer can connect it to an external display. I'd like to know if this issue is fixable through some firmware updates. Or should I buy a new machine?

I have a similar issue but I'm waiting my SOIC clips for more than a month now (delivery issue) to be sure. 

Meanwhile, I bought an `usb-c/HDMI dongle` and I can now have a second screen: https://www.cdiscount.com/informatique/clavier-souris-webcam/hub-usb-c-adaptateur-multiport-10-en-1-vers-hdmi-e/f-1070229-elc0713325730599.html#desc (i can't find an english link but you should get the idea)

Hoping this `workaround` can help you.

**Comment by jiahe224** [comment]
I installed the Thunderbolt Firmware Update Utility (https://pcsupport.lenovo.com/us/en/products/LAPTOPS-AND-NETBOOKS/THINKPAD-X-SERIES-LAPTOPS/THINKPAD-X1-CARBON-6TH-GEN-TYPE- 20KH-20KG/downloads/DS502615), 
but the NVM firmware version is 30 instead of 46, my problem is, I set the display to turn off after 10 minutes and the computer never goes to sleep, when the display turn off, the external display will sleep（I'm not sure that whether the status is sleep） because it loss of input signal。When pressed the keyboard  or moved the mouse   to light up the screen, the X1C will continue to emit a prompt sound that the USB device is not recognized, but the external display can display the image normally, and there is no yellow exclamation mark in the device manager(but the device list will be refreshed constantly), and the prompt sound will disappear after plugging and unplugging the display . My computer is connected to the display through Thunderbolt, and the external display can provide 65W power supply to the computer, I am not using an HDMI cable. I am not sure whether this problem is related to thunderbolt, because my other computer X13 is not installed with thunderbolt software, But everything works fine.

**Comment by atoch419** [comment]
Has anyone made any progress on this. I have exactly the same issue No thunderbolt no alt DP no hdmi no external display outputs showing up on Xrandr but fast charging still works. from reading this whole thread it seems like something is missing and the reported issue with a bad firmware isn't the whole story. Sounds like replacing the eporom and reflashing dosent work? could the issue be further down stream or are people thinking it's a signature issue? Interestingly I got my machine in mid 2020. I did all the updates on the Lenovo website in windows befor swapping to Linux so I should of installed the thunderbolt firmware fix. although I can't recall reading about this issue until it happened to me. could it be that damage was already done to the chip and even with new firmware it's life time shortened and this failure was going to happen sooner or later?

**Comment by RAY5D** [comment]
> Has anyone made any progress on this. I have exactly the same issue No thunderbolt no alt DP no hdmi no external display outputs showing up on Xrandr but fast charging still works. from reading this whole thread it seems like something is missing and the reported issue with a bad firmware isn't the whole story. Sounds like replacing the eporom and reflashing dosent work? could the issue be further down stream or are people thinking it's a signature issue? Interestingly I got my machine in mid 2020. I did all the updates on the Lenovo website in windows befor swapping to Linux so I should of installed the thunderbolt firmware fix. although I can't recall reading about this issue until it happened to me. could it be that damage was already done to the chip and even with new firmware it's life time shortened and this failure was going to happen sooner or later?

I still think the issue is with the firmware flash chip. The reason mine wasn't fixed is that the original chip has lost all the data, which seemed to be somewhat unique to each computer. I also had a Thinkpad tablet that is vulnerable to this problem, but I backed up the original data, then replaced the chip and updated to the latest version. In this way, no further damage can be done during the update process, and it's still working fine.

**Comment by fowoz** [comment]
> > Has anyone made any progress on this. I have exactly the same issue No thunderbolt no alt DP no hdmi no external display outputs showing up on Xrandr but fast charging still works. from reading this whole thread it seems like something is missing and the reported issue with a bad firmware isn't the whole story. Sounds like replacing the eporom and reflashing dosent work? could the issue be further down stream or are people thinking it's a signature issue? Interestingly I got my machine in mid 2020. I did all the updates on the Lenovo website in windows befor swapping to Linux so I should of installed the thunderbolt firmware fix. although I can't recall reading about this issue until it happened to me. could it be that damage was already done to the chip and even with new firmware it's life time shortened and this failure was going to happen sooner or later?
> 
> I still think the issue is with the firmware flash chip. The reason mine wasn't fixed is that the original chip has lost all the data, which seemed to be somewhat unique to each computer. I also had a Thinkpad tablet that is vulnerable to this problem, but I backed up the original data, then replaced the chip and updated to the latest version. In this way, no further damage can be done during the update process, and it's still working fine. 

![Screenshot_20230815_070105_com.google.android.apps.docs.jpg](https://github.com/fwupd/firmware-lenovo/assets/8021618/29dedcce-2a76-4c56-a2fe-b4edc5f6c308)

![Screenshot_20230815_070544_com.google.android.apps.docs.jpg](https://github.com/fwupd/firmware-lenovo/assets/8021618/8f5cbd5f-0792-4aea-86a2-3ba71f9a2230)

![Screenshot_20230815_070525_com.google.android.apps.docs.jpg](https://github.com/fwupd/firmware-lenovo/assets/8021618/c3391794-f497-4485-ab08-3918e67e7e30)

![Screenshot_20230815_070701_com.google.android.apps.docs.jpg](https://github.com/fwupd/firmware-lenovo/assets/8021618/1bbce57c-206c-450d-bbd4-3cc92c759516)

**Comment by jiahe224** [comment]
> I installed the Thunderbolt Firmware Update Utility (https://pcsupport.lenovo.com/us/en/products/LAPTOPS-AND-NETBOOKS/THINKPAD-X-SERIES-LAPTOPS/THINKPAD-X1-CARBON-6TH-GEN-TYPE- 20KH-20KG/downloads/DS502615), but the NVM firmware version is 30 instead of 46, my problem is, I set the display to turn off after 10 minutes and the computer never goes to sleep, when the display turn off, the external display will sleep（I'm not sure that whether the status is sleep） because it loss of input signal。When pressed the keyboard or moved the mouse to light up the screen, the X1C will continue to emit a prompt sound that the USB device is not recognized, but the external display can display the image normally, and there is no yellow exclamation mark in the device manager(but the device list will be refreshed constantly), and the prompt sound will disappear after plugging and unplugging the display . My computer is connected to the display through Thunderbolt, and the external display can provide 65W power supply to the computer, I am not using an HDMI cable. I am not sure whether this problem is related to thunderbolt, because my other computer X13 is not installed with thunderbolt software, But everything works fine.

I no longer use the monitor to power the computer, but use the power supply that comes with the computer, and use a type-c cable to connect the thunderbolt interface and the monitor, and it seems to be normal now.

**Comment by atoch419** [comment]
Hey REY5D So i think im in exactly the same boat as you now
I checked my eeprom and it is full of fs if I flash anything to it nothing changes. if use hex dump it there is absolutely nothing there. Can you remind me if you tried the zero filling and booting after replacing the chip befor flashing the firmware. I dont understand how the person at this link https://nadim.computer/posts/2018-10-26-thinkpad.html which has been posted befor could of flashed the chip with zeros then flashed the firmware and got a fully working machine if their has to be something on the chip that ids it. unless the chips ID or serial number or somthing is stored in the chip and recorded else where in the machine (maybe on the other bios chip?) or another chip?

**Comment by RAY5D** [comment]
> Hey REY5D So i think im in exactly the same boat as you now I checked my eeprom and it is full of fs if I flash anything to it nothing changes. if use hex dump it there is absolutely nothing there. Can you remind me if you tried the zero filling and booting after replacing the chip befor flashing the firmware. I dont understand how the person at this link https://nadim.computer/posts/2018-10-26-thinkpad.html which has been posted befor could of flashed the chip with zeros then flashed the firmware and got a fully working machine if their has to be something on the chip that ids it. unless the chips ID or serial number or somthing is stored in the chip and recorded else where in the machine (maybe on the other bios chip?) or another chip?

I didn't zero the EEPROM out before all the attempts of re-flashing, I did it for at least some of the binaries, but these attempts were apparently not successful. 
I did some more searching and found the schematic of the motherboard, the "model" of which is NM-B48 on my laptop. The flash chip, U184, is connected only to U182 and U192, U182 is a JHL6540 Thunderbolt controller, and U192 is a TPS65988 PD CONTROLLER. I think only the TB controller has the processing capability to run a firmware since it has a XTAL Y7 nearby, while U192 looks mostly like a dumb power chip. 
Whether the JHL6540 contains a serial is unknown, it's also possible that the JHL6540 get the serial from somewhere else (maybe through communicating with BIOS?), but this is just a hypothesis. Given that people have had success with other Thinkpad models, I'm starting to doubt that the firmware contain unique data for each chip at all. I'll try the flashing process more carefully and see what happens. 
I'm attaching the schematic and relevant photos below. I also found [this post from the badcaps forum](https://badcaps.net/forum/showthread.php?p=1162221), which looks like the exact issue we're having, although they are struggling with the binary as well. 

![20230407_023418](https://github.com/fwupd/firmware-lenovo/assets/12710440/34f6ba91-5828-4cf8-95e5-fd57d93c41bb)

![20230407_023344](https://github.com/fwupd/firmware-lenovo/assets/12710440/aac5710f-0599-4b46-9e49-0cd0062d59ab)

[Lenovo X1 Carbon LCFC YODA-2 NM-B48 6th Gen.pdf](https://github.com/fwupd/firmware-lenovo/files/12615486/Lenovo.X1.Carbon.LCFC.YODA-2.NM-B48.6th.Gen.pdf)

**Comment by dal00** [comment]
Hey! That's my original post but I've kind of given up on this. I did swap the chip even and it didn't help. So must be the firmware. I'll follow your progress though.

**Comment by atoch419** [comment]
things don't fully add up. with other thinkpads people have fixed this issue flashing the official .bin from Lenovo after over writing with zeros. so it dosent seem like there should be something in the binary that is IDing the chip. You your self have managed to replace a chip on another machine with a working binary pulled from the same machine so it dosent make sense that the chip ID is baked in to silicone either as then the replacement chip wouldn't have worked.
as an experiment could you back up the chip on your other working machine. then flash the official Lenovo binary on to it and see what happens? maybe you don't want to mess with a working machine though

**Comment by RAY5D** [comment]
> things don't fully add up. with other thinkpads people have fixed this issue flashing the official .bin from Lenovo after over writing with zeros. so it dosent seem like there should be something in the binary that is IDing the chip. You your self have managed to replace a chip on another machine with a working binary pulled from the same machine so it dosent make sense that the chip ID is baked in to silicone either as then the replacement chip wouldn't have worked. as an experiment could you back up the chip on your other working machine. then flash the official Lenovo binary on to it and see what happens? maybe you don't want to mess with a working machine though

The one I replaced was flashed with its own extracted binary, so it for sure will work. I did it because I think the software update will damage the flash IC to some extent, and wanted to be safe with a new IC before trying the software update. The experiment idea you had is actually pretty smart, but I'm afraid it's not practical on my side because the model (X1 Tablet 3rd Gen) requires desoldering the flash chip to read and write it correctly (imagine messing with the board twice just to try one binary, rip). But I'll try something else.

**Comment by atoch419** [comment]
Ok bit more digging today
turns out i was using my clip wrong or the software i was using wasn't great, anyway i have swapped to flashrom  and got everything working. pulled the original bin off stuck a zero file on booted up stuck the TBT.bin on. got my fast charging back but no display outputs. did some reading of the thunder spy presentation you posted and noticed something. the strings after RSA match up between both firmware but the section after SEC DGST is different on my pulled firmware and on the TBT.bin did a search for SEC DGST and it is something to do with keys so will look at transplanting this section from my dump in to the TBT.bin and try re flashing it tomorrow

**Comment by LinuxOnTheDesktop** [comment]
I have an X1CG6. I find that at some point the HDM port stopped working. I was about to try to get my motherboard fixed or replaced . . and I find now that one of Lenovo's carefully crafted updates might have borked the port.

I seem to see too - from all the posts above - that . .  there is no fix, at least not for Linux.

I am hoping that someone can tell that actually there is a fix - and one that does not involved some special piece of flashing hardware.

It does not help that finding out which version of the relevant firmware I am running does not seem very easy on fwupd.

**Comment by LinuxOnTheDesktop** [comment]
@mrhpearson 

I bought my expensive X1CG6 in order partly to give presentations. Seemingly a firmware update has broken the HDMI port (and perhaps - for something has broken this - the main charging port). I bought a USB-C-to-HDMI adaptor to try to work around the problem. It does not work. So now I cannot give presentations. How is this Lenovo going to make this situation right?

**Comment by mrhpearson** [comment]
I think you need to contact customer support for this issue
Mark

**Comment by LinuxOnTheDesktop** [comment]
@mrhpearson 

Thanks, but I fear I'd have terrible trouble getting someone who is able and willing to help (and, perhaps, who has even heard of Linux). Could you supply an email address to which to write? Please note that, having had terrible problems ever since I got this machine, if I replace the machine it might not be with a ThinkPad. The problems: the trackpoint have never worked with Linux, despite a motherboard change; the motherboard change means one firmware update will never install, and now this broken port.

**Comment by mrhpearson** [comment]
As this is a FW update problem and you should be able to demonstrate HDMI not working from the BIOS setup menu. I don't think the OS should matter.
If the system was brought with Windows they may ask you to boot that to confirm, but that shouldn't be a big deal.
Afraid I can't help with RMA's - I'm not in the support organisation. I recommend calling them.

**Comment by LinuxOnTheDesktop** [comment]
@mrhpearson 

I emailed Lenovo support, giving a brief account of the issue. The response: 'We regret to inform you that the repair of your machine is not covered under the terms of warranty. Please advise if you wish to proceed with a billable repair.' (I know that you do not work in support, but whom else are we to turn to?) I will try reasoning with the support people . .

UPDATE: the support people were unable or unwilling to discuss anything with me; they kept repeating that I would have to pay.

**Comment by NightWinkle** [comment]
Just as a follow up for anyone with the same issue : 
I tried resoldering another chip and flashing the TBT firmware and it didn't work.

**Comment by theajmalrazaq** [comment]
Any leads, how can i fix this problem??

**Comment by ymilly** [comment]
Similar to many in this thread. I bought a used X1 G6 in mid 2023. It came with Windows installed. I installed Ubuntu and video out worked (USB and HDMI).

As it was a new install, I was running through the updates shown to me in Ubuntu. One was for a firmware. My external monitor was plugged in and working (via USB C). Upon reboot, the external monitor was not receiving video, and I've never been able to get video output to work again via USB C or HDMI in Windows or Linux.

I'm not sure what the exact connection of steps were that caused the video output to brick. Was it because I was running an external monitor while updating the firmware? Did I update the firmware out of order? Was the wrong firmware detected? Did updating firmware in Linux, when previous updates had been done via Windows, creating inconsistencies?

Lenovo support has been total hogwash. After being a long time Lenovo advocate, I've fallen into the camp of never buying another, and telling others the same.

But I'd like to understand what happened, and if there are options to get it working. It seems so hit and miss to troubleshoot this issue. No clear answers from anything I've found online.

**Comment by latonita** [comment]
x1c6, same issue. actually everything was working fine until I installed ubu24 from scratch and it did some updates :(
Now neither hdmi output nor usb-c hub  with hdmi do not work, I have no way to use external monitor anymore...
bios 1.64 / ec 1.22
I didn't get how to do update thunderbolt fw with this tool?

**Comment by jclaveau** [comment]
> x1c6, same issue. actually everything was working fine until I installed ubu24 from scratch and it did some updates :( Now neither hdmi output nor usb-c hub with hdmi do not work, I have no way to use external monitor anymore... bios 1.64 / ec 1.22 I didn't get how to do update thunderbolt fw with this tool?

I'm on Ubuntu also and, even if my hdmi port is bricked, I can use hdmi through this usb-c dongle https://www.cdiscount.com/informatique/clavier-souris-webcam/hub-usb-c-adaptateur-multiport-10-en-1-vers-hdmi-e/f-1070229-elc0713325730599.html

I guess I'm lucky

To avoid you digging into a french page, here is the ref:
- Brand: AYYBF
- Device: HUB USB C, Adaptateur 10 en 1 avec 4K-HDMI, VGA, USB 3.0, Type C, Port PD, Ethernet RJ45, Lecture Carte SD/TF, AUX 3,5 mm Audio

Btw, I also have a HP Port Replicator plugged on the USB-A port so I can use 2 displays

**Comment by latonita** [comment]
I tried many ways - reflashing TBT chip with different donor firmwares, but no luck. I need to use external monitor.
I suppose most of adapters usb-c to hdmi will not work, since they just transport displayport. I tried couple usb-c docks - no luck.
And I need something like external card. Which is usb2>hdmi? Any recommendations?
I read about DisplayLink, but those I saw are $100+ which is too much for me. Any solution around 30-40?

**Comment by latonita** [comment]
Just to add a bit to worlds' entropy...  I tried to erase SPI flash, then reboot to windows, then flash TBT, then hold reset button,, remove battery, remove CMOS battery, reboot.... 
Diagnostic tool shows TBT version (latest which i flashed to SPI)...
However, no good results. hdmi/usb-c hdmi dock not working. Windows thunderbolt About doesnt show controller version.

![image](https://github.com/user-attachments/assets/8a6b7342-e4e6-4973-b599-c44eca566319)

**Comment by LordChaos73** [comment]
![IMG_20241114_122426990](https://github.com/user-attachments/assets/269cc02f-8f58-42d2-b01e-ef3f1f2a095a)

**Comment by vava8602** [comment]
Its 2024 and now I have the same problem, couldnt connect to external monitor whatsoever. Any luck by anybody?

**Comment by freakzlike** [comment]
I have two Lenovo notebooks (L15 and T15p). Upgraded my Thunderbolt 4 to Firmware 1.0.18 with the L15 and it was working a couple of days without issues. Switch to my T15p and the HDMI Monitor is no longer working, like it is not connected. Switched back to the L15 and no longer working there too. I can't get any monitor connection working now with my Thunderbolt 4.

Tested:
* Other Monitor
* Other HDMI cable
* Both HDMI ports
* DP port

No monitor connection possible whatsoever

**Comment by mrhpearson** [comment]
Hi @freakzlike - I think this is a different problem (as it's related to your dock FW).

Could you create a new ticket please? I'll get an internal ticket open for tracking and we can look into it. 
It would also be useful to get a kernel log to review - can you send to mpearson-lenovo at squebb dot ca please?

Thanks
Mark

**Comment by maxvelkir** [comment]
Has this been resolved or moved to a new issue?

Got an x1gen6 and no external displays work, tried to connect through DP to USB-C and HDMI. `fwupdmgr` shows no updates pending, everything on latest version. `boltctl list` is empty. Went as far as installing win 10 to install the critical thunderbolt firmware but no luck.

**Comment by mrhpearson** [comment]
It's been a while - but as I recall the FW was released (24W/1.47).
If you're not seeing a TBT device it's possible your controller is dead I'm afraid. You'll need to contact Lenovo support as it's a HW issue

**Comment by latonita** [comment]
I even tried to desolder flash chip, flash it with different firmware versions and then solder back - nothing helps...

**Comment by jiahe224** [comment]
> > I installed the Thunderbolt Firmware Update Utility (https://pcsupport.lenovo.com/us/en/products/LAPTOPS-AND-NETBOOKS/THINKPAD-X-SERIES-LAPTOPS/THINKPAD-X1-CARBON-6TH-GEN-TYPE- 20KH-20KG/downloads/DS502615), but the NVM firmware version is 30 instead of 46, my problem is, I set the display to turn off after 10 minutes and the computer never goes to sleep, when the display turn off, the external display will sleep（I'm not sure that whether the status is sleep） because it loss of input signal。When pressed the keyboard or moved the mouse to light up the screen, the X1C will continue to emit a prompt sound that the USB device is not recognized, but the external display can display the image normally, and there is no yellow exclamation mark in the device manager(but the device list will be refreshed constantly), and the prompt sound will disappear after plugging and unplugging the display . My computer is connected to the display through Thunderbolt, and the external display can provide 65W power supply to the computer, I am not using an HDMI cable. I am not sure whether this problem is related to thunderbolt, because my other computer X13 is not installed with thunderbolt software, But everything works fine.
> 
> I no longer use the monitor to power the computer, but use the power supply that comes with the computer, and use a type-c cable to connect the thunderbolt interface and the monitor, and it seems to be normal now.

Damn it, everything was working fine. When I brought my laptop home from work and connected it to the monitor, the system alerted me about an “unrecognized USB device.” In Device Manager, under “Universal Serial Bus controllers,” there was an “Unknown USB device (Port reset failed)” with a yellow exclamation mark, but it disappeared after appearing once.

Out of desperation, I reinstalled the system. Thunderbolt software shows everything as normal, NVM version is 47, all related services are running, and I even modified the BIOS settings following this article (https://zhuanlan.zhihu.com/p/529323703), but the problem persists!

【update】The monitor's power cable came loose 😓. After plugging it back in, everything seemed to work normally again. However, when I used the monitor to power the computer via the USB-C cable while also transmitting video and audio, the issue resurfaced. At this point, I need to turn off the monitor, disconnect both its power cable and the USB-C cable connected to the computer. Then, I shut down the computer, reconnect both the computer's and monitor's power cables, plug in the USB-C cable between them, turn on the computer, and finally power on the monitor.

Someone replaced the JHL6540 chip for $2 and everything worked fine.

**Comment by dal00** [comment]
> Someone replaced the JHL6540 chip for $2 and everything worked fine.

Do you have a link to an article describing this procedure?

**Comment by jiahe224** [comment]
> > Someone replaced the JHL6540 chip for $2 and everything worked fine.
> 
> Do you have a link to an article describing this procedure?

 this article (https://zhuanlan.zhihu.com/p/529323703), Comments section, You can use your browser's translation feature.

**Comment by RAY5D** [comment]
OK, I did the chip replacement. It did something, but I'm not sure if it's a complete fix. 

![Image](https://github.com/user-attachments/assets/87db482a-8bc6-4485-b54c-12ad8875f31d)

HDMI is fixed. 
Charging port negotiation works, can get 20V. 
Then I tried a TB4 dock, model DK2131, type 40B0. 
The power button on the dock can turn the computer on/off. 
Nothing else works, no dock USB, etc. 
Thunderbolt software doesn't show the dock as a connected device, or the Thunderbolt ports 0 and 1. 
Thunderbolt controller doesn't show up in Device manager. 

I'm not sure if there is a compatibility issue with the dock, but this is currently the only Thunderbolt device I can test with. Does anyone remember if their Thunderbolt software shows the ports when nothing is connected when their TB controller still worked? I think it should? 

<img width="1187" height="1411" alt="Image" src="https://github.com/user-attachments/assets/cdab7ed1-35e5-4c8a-914a-ef16c399e019" />

I think the firmware binary has a role in this. My suspicion is that the TB controller JHL6540 loads firmware from external flash at power up, and keep the data in some sort of internal flash to execute. That's why just replacing the external flash chip won't fix the problem, the internal memory of the JHL6540 was damaged due to the bug in the firmware. (It also explains why filling the external flash with 0x00 is needed before flashing the actual binary, as it triggers a load on the controller.) I tried a couple of binaries I found a good while ago when I first attempted to fix the issue, but the best I could achieve is working HDMI, working dock power button, basically what I described above. Or this is already good on the firmware side, and the issue is with drivers on the OS? 

There are some other methods described in the zhihu post. But I think I've broken the IC again by flashing random binaries... so don't do that. I'll probably replace it again and try these methods.

**Comment by RAY5D** [comment]
OK, if you still haven't thrown away the computer after all these years, then good news! I realized I made a silly mistake in my previous attempt, and the IC replacement should work completely. What I found was that Thunderbolt 3 was somehow set to Disabled in BIOS-Secutiry-I/O Port Access, and by setting it back to Enabled, the Thunderbolt controller showed up in Device Manager, Thunderbolt Software, and the dock fully functions. 
So to summarize, the fix requires replacing the SOP-8 flash IC, flashing a binary into the new IC, and replacing the JHL6540 Thunderbolt Controller, which is an under-filled BGA. The flash IC is Winbond W25Q80DVSNIG, the JHL6540 can be found on Ebay/Aliexpress, and I think rossmanngroup also sells it. You'll need a flash programmer (random Chinese ones with CH341 is very common), and SOP-8 clips as well. 
As for the binary, maybe you can try extracting the existing content of the damaged IC and use that as the replacement. It should be a safe bet without downloading suspicious executable from the Internet. If that doesn't work, the binary that worked for my computer was called "tbt_20220530 from reddit.bin" in my folder of random binaries. It appears that the original uploader has deleted his Reddit post, so I don't know who to give credit to. Also, it seems like sharing binary is not allowed here anymore? Not sure if I'll get banned or something, but I'll not post it right away until getting permission from mods and interest from others. After all, this is a pretty involved fix, and likely requires bringing the computer to a local repair professional that can do BGA rework for you. 

BIOS setting that I missed: 
<img width="4032" height="3024" alt="Image" src="https://github.com/user-attachments/assets/bcd4c0ba-0473-40cc-a788-f28f25847ddb" />

Thunderbolt Software after I turned it back on: 
<img width="4032" height="3024" alt="Image" src="https://github.com/user-attachments/assets/a7c501fa-c9b9-4f2b-bbc8-51294b8cf5b5" />

Now working with dock: 
<img width="4032" height="3024" alt="Image" src="https://github.com/user-attachments/assets/569b60a8-0d2f-44fc-a8ec-6b4800d9ccef" />
